### DISCUSSION TITLE
nx version for web apps (non-package.json)

### DISCUSSION THREAD

**Discussion by kunyan** [nrwl/nx#23386] [Feature Requests]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description

Since `nx release` more and more stable, I just want to ask how to use nx release manage web apps

As you know, all apps in nx monorepo do not have their own package.json. so there are no place for `nx release` to bump version.

Although we would not publish web apps to npm registry in most of time, we still need the git tag, changelog to identify the changes/release log for a web app

**Comment by fahslaj** [comment]
Hi @kunyan , `nx release`'s main operations are divided between three different subcommands that can be used directly.

If you are just using `nx release` for applications, you can add those apps to the `release.projects` array in nx.json, then use the `nx release changelog` subcommand to just perform the changelog generation and git commit and tag step of the process. Alternatively, you can use the changelog programmatic API to call it directly (https://nx.dev/features/manage-releases#using-the-programmatic-api-for-nx-release).

**Comment by laibold** [comment]
Can you give an example on how to use `nx release` for apps without a `package.json`? With every subcommand of  `nx release`, I get the error message 

```
 >  NX   The project "example-app" does not have a package.json available at apps/example-app/package.json.

To fix this you will either need to add a package.json file at that location, or configure "release" within your nx.json to exclude "example-app" from the current release group, or amend the packageRoot configuration to point to where the package.json should be.
```

**Comment by kunyan** [comment]
The only workaround I found is created a `package.json` for each app. with the app name and version. then `nx release` will work as expected

The `package.json` only for bump version, not include any `dependencies` or `devDependencies`

**Comment by ddspiff** [comment]
after being stuck on this for a while, i think this is the proper way to do it until someone tell me otherwise.  in `nx.json`, i added:
```
  "generators": {
    "@nx/js": {
      "release-version": {
        "packageRoot": "."
      }
    },
   ```
and `npx nx release version --dry-run --first-release` seemed to be working fine.

**Reply by antoniobfm** [reply]
Solved for me!
Thanks for sharing @ddspiff

**Reply by mroobert** [reply]
Can you please explain a little bit why did you add this since in the docs https://nx.dev/nx-api/js/generators/release-version for **_packageRoot_** we have this description: _The root directory of the directory (containing a manifest file at its root) to publish. **Defaults to the project root**_

**Reply by vzsoares** [reply]
@mroobert `.` will resolve to where nx.json is, regardless of project.

**Reply by vzsoares** [reply]
> after being stuck on this for a while, i think this is the proper way to do it until someone tell me otherwise. in `nx.json`, i added:
> 
> ```
>   "generators": {
>     "@nx/js": {
>       "release-version": {
>         "packageRoot": "."
>       }
>     },
> ```
> 
> and `npx nx release version --dry-run --first-release` seemed to be working fine.

this makes my version to be bumped times the amount of projects i have. So if i have 5 projects a major release will go from 1.0.0 to 6.0.0 ):

**Comment by vzsoares** [comment]
I ended up adding a `project.json` in the root of the repository and adding only it in nx.json `"projects": ["rootProject"]`.

nx.json
```
    "release": {
        "projects": ["rootProject"]
    }
```
project.json
```
{
    "name": "rootProject",
    "root": ".",
...
```


- [root project.json](https://nx.dev/features/run-tasks#run-rootlevel-tasks)
- [configure projects](https://nx.dev/features/manage-releases#customizing-releases)

**Comment by ak274** [comment]
@JamesHenry @fahslaj  

 I have configured: ```"projects": ["apps/*"]``` like this.
and now if i make a change in a lib (non0-buildable) which is dependency of only one app.. I am expecting it will version bump only the app which is dependent but it updating all the apps in the folder. I want it to work similar how affected build target works.
